snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,315 @@
+	}
+
+	totalKnownLastBlockStakePercent := float64(netInfo.totalWeightWeKnowItsLastAcceptedBlock) / float64(netInfo.totalValidatorWeight)
+	stakeAheadOfUs := float64(netInfo.totalPendingStake) / float64(netInfo.totalValidatorWeight)
+
+	// Ensure we have collected last accepted blocks for at least 80% (or so) stake of the total weight we are connected to.
+	if totalKnownLastBlockStakePercent < minimumStakeThresholdRequiredForNetworkInfo {
+		s.log.Trace("Not collected enough information about last accepted blocks for the validators we are connected to",
+			zap.Float64("ratio", totalKnownLastBlockStakePercent))
+		return false
+	}
+
+	if stakeAheadOfUs < knownStakeThresholdRequiredForAnalysis {
+		s.log.Trace("Most stake we're connected to has the same height as we do",
+			zap.Float64("ratio", stakeAheadOfUs))
+		return false
+	}
+
+	return true
+}
+
+func (s *snapshotter) getNetworkSnapshot() (snapshot, bool) {
+	ourLastAcceptedBlock := s.lastAccepted()
+
+	netInfo, err := s.getNetworkInfo(ourLastAcceptedBlock)
+	if err != nil {
+		return snapshot{}, false
+	}
+
+	if !s.validateNetInfo(netInfo) {
+		return snapshot{}, false
+	}
+
+	snap := snapshot{
+		lastAcceptedBlockID:  netInfo.nodeWeightToLastAccepted,
+		totalValidatorWeight: netInfo.totalValidatorWeight,
+	}
+
+	return snap, true
+}
+
+func (s *snapshotter) getNetworkInfo(ourLastAcceptedBlock ids.ID) (netInfo, error) {
+	var res netInfo
+
+	validators := nodeWeights(s.connectedValidators().List())
+	nodeWeightToLastAccepted := make(nodeWeightsToBlocks, len(validators))
+
+	for _, vdr := range validators {
+		lastAccepted, ok := s.lastAcceptedByNodeID(vdr.ID)
+		if !ok {
+			continue
+		}
+		nodeWeightToLastAccepted[vdr] = lastAccepted
+	}
+
+	totalValidatorWeight, err := validators.totalWeight()
+	if err != nil {
+		s.log.Error("Failed computing total weight", zap.Error(err))
+		return netInfo{}, err
+	}
+	res.totalValidatorWeight = totalValidatorWeight
+
+	totalWeightWeKnowItsLastAcceptedBlock, err := nodeWeightToLastAccepted.totalWeight()
+	if err != nil {
+		s.log.Error("Failed computing total weight", zap.Error(err))
+		return netInfo{}, err
+	}
+	res.totalWeightWeKnowItsLastAcceptedBlock = totalWeightWeKnowItsLastAcceptedBlock
+
+	prevLastAcceptedCount := len(nodeWeightToLastAccepted)
+
+	// Ensure we have collected last accepted blocks that are not our own last accepted block.
+	nodeWeightToLastAccepted = nodeWeightToLastAccepted.filter(func(id ids.ID) bool {
+		return ourLastAcceptedBlock.Compare(id) != 0
+	})
+
+	res.nodeWeightToLastAccepted = nodeWeightToLastAccepted
+
+	totalPendingStake, err := nodeWeightToLastAccepted.totalWeight()
+	if err != nil {
+		s.log.Error("Failed computing total weight", zap.Error(err))
+		return netInfo{}, err
+	}
+
+	res.totalPendingStake = totalPendingStake
+
+	s.log.Trace("Excluding nodes with our own height", zap.Int("prev", prevLastAcceptedCount), zap.Uint64("new", totalPendingStake))
+
+	return res, nil
+}
+
+type netInfo struct {
```

Comment:
Could we include some comments explaining the phases of filtering?

The process seems to be:
1. Gather the last accepted block from each validator
2. Filter out any blocks that match our own last accepted block in `getNetworkInfo`
3. Filter out any blocks that we have marked as processing in `areWeBehindTheRest`

but this requires reading from the bottom to top of the file to understand the flow.

How do we handle blocks that are accepted by the network, but we have not added to processing yet because we have not fetched the full ancestry?

Since we are using `consensus.Processing(...)`, we'll only count blocks that we have already verified, but I thought the intent behind this PR was to detect scenarios where we are sufficiently far behind that the blocks reported as the last accepted block by our peers are far enough ahead that we have not verified them.